dali/kernels/imgproc/resample/resampling_impl.cuh

```diff
@@ -0,0 +1,280 @@
+
+        for (int c = 0; c < channels; c++)
+          out_row[channels * dx + c] = clamp<Dst>(tmp[c] * norm);
+      }
+    }
+  }
+}
+
+/// @brief Implements vertical resampling for a custom ROI
+/// @param x0 - start column, in output coordinates
+/// @param x1 - end column (exclusive), in output coordinates
+/// @param y0 - start row
+/// @param y1 - end row (exclusive)
+/// @tparam static_channels - number of channels, if known at compile time
+///
+/// The function fills the output in block-sized horizontal spans.
+/// Filter coefficients are pre-calculated for each horizontal span to avoid
+/// recalculating them for each column, and stored in a shared memory block.
+template <int static_channels = -1, typename Dst, typename Src>
+__device__ void ResampleVert_Channels(
+    int x0, int x1, int y0, int y1,
+    float src_y0, float scale,
+    Dst *out, int out_stride,
+    const Src *in, int in_stride, int in_h, int dynamic_channels,
+    ResamplingFilter filter, int support) {
+  using resample_shared::coeffs;
+
+  src_y0 += 0.5f * scale - 0.5f - filter.anchor;
+
+  const float filter_step = filter.scale;
+
+  const int coeff_base = support*threadIdx.y;
+  const int channels = static_channels < 0 ? dynamic_channels : static_channels;
+
+  const float bias = std::is_integral<Dst>::value ? 0.5f : 0;
+
+  for (int i = y0; i < y1; i+=blockDim.y) {
+    int dy = i + threadIdx.y;
+    const float sy0f = dy * scale + src_y0;
+    const int sy0 = ceilf(sy0f);
+    float f = (sy0 - sy0f) * filter_step;
+    __syncthreads();
+    for (int k = threadIdx.x; k < support; k += blockDim.x) {
+      float flt = filter(f + k*filter_step);
+      coeffs[coeff_base + k] = flt;
+    }
+    __syncthreads();
+
+    if (dy >= y1)
+      continue;
+
+    Dst *out_row = &out[dy * out_stride];
+
+    float norm = 0;
+    for (int k = 0; k < support; k++) {
+      norm += coeffs[coeff_base + k];
+    }
+    norm = 1.0f / norm;
+
+    for (int j = x0 + threadIdx.x; j < x1; j += blockDim.x) {
+      Dst *out_col = &out_row[j * channels];
+      const Src *in_col = &in[j * channels];
+
+      if (static_channels < 0) {
+        for (int c = 0; c < channels; c++) {
+          float tmp = bias;
+
+          for (int k = 0; k < support; k++) {
+            int y = sy0 + k;
+            int ysample = y < 0 ? 0 : y >= in_h-1 ? in_h-1 : y;
+            float flt = coeffs[coeff_base + k];
+            Src px = __ldg(in_col + in_stride * ysample + c);
+            tmp += px * flt;
+          }
+
+          out_col[c] = clamp<Dst>(tmp * norm);
+        }
+      } else {
+        float tmp[static_channels < 0 ? 1 : static_channels];  // NOLINT - not a variable length array
+        for (int c = 0; c < channels; c++)
+          tmp[c] = bias;
+
+        for (int k = 0; k < support; k++) {
+          int y = sy0 + k;
+          int ysample = y < 0 ? 0 : y >= in_h-1 ? in_h-1 : y;
+          float flt = coeffs[coeff_base + k];
+          for (int c = 0; c < channels; c++) {
```

Comment:
I wonder if loop unrolling would help anyhow for channels.

Reply:
I think compiler can pick up, that channels depends on template parameter and unrolls it automagically, if appropriate. I've checked it in compiler explorer on similar but simpler can and it did do exactly that.